### PR TITLE
[flang] Add structure constructor with allocatable component

### PR DESCRIPTION
**Pull request by kkwli**
This PR is to enable the structure constructor with allocatable component support. Handling of `null()` for the allocatable component is added.

A separate PR will be created for enabling the related test cases in llvm-test-suite.

### CODE REVIEW

File: flang/lib/Evaluate/check-expression.cpp [hunk 1]
```diff
@@ -188,6 +188,12 @@ struct IsActuallyConstantHelper {
     return common::visit([this](const auto &y) { return (*this)(y); }, x.u);
   }
   bool operator()(const StructureConstructor &x) {
+    // If the comp-spec of a derived-type-spec is a structure constructor that
```
Comment (klausler) [inline flang/lib/Evaluate/check-expression.cpp:191]: "derived-type-spec" is a grammar production associated with types.  I don't know how that can be relevant here.  And I think that `UnwrapExpr<Constant<SomeDerived>>(x)` will always be null -- if the subexpression is a structure constructor, it's not a `Constant` -- so the effect is to just return `false` for all structure constructors.  So I don't understand this part at all.  What are you trying to accomplish here?

Comment (kkwli) [reply flang/lib/Evaluate/check-expression.cpp:191]: Thanks for the quick review. I was trying to handle the following case. My original approach is not handling it correctly. I will update the patch with the better approach shortly
```fortran
type dt0
  integer, allocatable :: a(:)
end type
type dt1
  type(dt0) :: b
end type
type(dt1) xdt1 = dt1(dt0(5))
```

**Review by jeanPerier — COMMENTED** [review]
Thanks for working on this!

File: flang/lib/Lower/ConvertConstant.cpp [hunk 2]
```diff
@@ -362,8 +363,18 @@ static mlir::Value genStructureComponentInit(
       loc, fieldTy, name, recTy,
       /*typeParams=*/mlir::ValueRange{} /*TODO*/);
 
-  if (Fortran::semantics::IsAllocatable(sym))
-    TODO(loc, "allocatable component in structure constructor");
+  if (Fortran::semantics::IsAllocatable(sym)) {
+    if (Fortran::evaluate::UnwrapExpr<Fortran::evaluate::NullPointer>(expr)) {
```
[the remark this answers is not in the document]
Comment (kkwli) [reply flang/lib/Lower/ConvertConstant.cpp]: Thanks for pointing it out. I will update it as you suggest.

**Review by jeanPerier — APPROVED** [review]
Thanks, LGTM

**Comment by kkwli** [comment]
Enable tests in llvm-test-suite - https://github.com/llvm/llvm-test-suite/pull/76

**Pull request merged**